### PR TITLE
3 get /api/topics

### PR DESCRIPTION
**Pull request by polinausatova**

### CODE REVIEW

File: __tests__/newsApp.test.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
+const app = require("../newsApp");
+const data = require("../db/data/test-data");
+const request = require("supertest");
+const db = require("../db/connection.js");
+const seed = require("../db/seeds/seed");
+
+beforeEach(() => {
+  // console.log('test seeding done');
```
Comment (corkesw) [inline __tests__/newsApp.test.js]: Don't forget to remove any commented out code or console.logs before creating your pull request

File: newsControllers/newsControllers [hunk 2]
```diff
@@ -0,0 +1,15 @@
+const {fetchTopics } = require('../newsModels/newsModels')
+
+
+exports.getTopics = (req, res) => {
```
Comment (corkesw) [inline newsControllers/newsControllers]: next is missing from your parameters so will be undefined in your code

File: newsModels/newsModels [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+const db = require("../db/connection.js");
+//const { updateDataReversed } = require("../db/utility.js");
+
+exports.fetchTopics = (url) => {
+    return db.query("SELECT * FROM topics").then((topics) => {
+        console.log('models OK');
+        // console.log(url);
+        // console.log(topics);
+       
+        const formattedTopics = topics.rows.map(({ slug, description }) => {
```
Comment (corkesw) [inline newsModels/newsModels:8]: The data should not require any formatting, you are effectively destructuring, then restructuring the objects here

File: newsControllers/errorHandlingControllers.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
+exports.handlePSQL400s = (error, request, response, next) => {
```
Comment (corkesw) [inline newsControllers/errorHandlingControllers.js]: There are no psql or custom error to handle in your app yet so this code should not be included in this PR.

File: newsApp.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
+const express = require('express');
+const app = express();
+
+const {
+    handle500Statuses,
+    handlePSQL400s,
+    handleCustomErrors,
+} = require('./newsControllers/errorHandlingControllers');
+
+app.use(express.json());
```
Comment (corkesw) [inline newsApp.js]: We do not need to parse any json in the app yet so this middleware should not yet be included

File: __tests__/newsApp.test.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        //Act
+    .get("/api/topics")
+        //Assert
+    .expect(200)
+    .then(({ body }) => {
+        expect(body).toHaveProperty('topics', expect.any(Array)); 
+        expect(body.topics.length).toBe(data.topicData.length);
+        
+        for (let i=0; i<body.topics.length; i++) {
+        expect(Object.keys(body.topics[i]).sort()).toEqual(['slug', 'description'].sort());
```
Comment (corkesw) [inline __tests__/newsApp.test.js]: We should also check that the values are of the type that we expect. Jest allowed us to use `any` syntax here, eg `expect.any(String)`

**Review by corkesw — COMMENTED** [review]
Hi Polina, good work so far. Please be sure to tidy up any comments and console.logs before submitting code. I've made some suggestions on the testing and highlighted some unneeded code in the model. You should also be careful not include code that we don't need yet. If you have any questions then give me a shout 🙂

File: __tests__/newsApp.test.js (hunk 1 above)
Comment (polinausatova) [reply __tests__/newsApp.test.js]: Is it absolutely not OK to leave any comments? Was thinking I might need some of these console.logs (not highlighted one) later for debugging the same types of errors with further use of same modules. I also leave some commented questions for myself so I can ask tutor or think one more time later.

File: newsControllers/newsControllers (hunk 2 above)
Comment (polinausatova) [reply newsControllers/newsControllers]: Yes, I got a 'next is undefined' message at some point - but it resolved itself after I made some other changes, so I did not know the reason. Thank you for highlighting.

File: newsModels/newsModels (hunk 3 above)
Comment (polinausatova) [reply newsModels/newsModels:8]: I thought this way was given us at some lecture with a reason I cannot remember now. Should I change this code, or is it just for future use?

File: newsControllers/errorHandlingControllers.js (hunk 4 above)
Comment (polinausatova) [reply newsControllers/errorHandlingControllers.js]: Was not sure yesterday how to do proper error handling for this ticket, so decided to make this rather than nothing, though it did look unnecessary. Everything makes more sense with today's lecture.

File: __tests__/newsApp.test.js (hunk 6 above)
Comment (polinausatova) [reply __tests__/newsApp.test.js]: Slightly irrelevant here, but it's about how I designed my tests: is it ok to use or copy the same algorithm from test to the utility used in the method? We can copy a bug here if the logic is not good and have a falsely good test outcome. Is this why just inserting manually some test values is preferable?

**Comment by polinausatova** [comment]
Thank you so much for your comments Stephen! I would like to include a proper test for error handling as Jim shown us today and make a new PR if you don't mind.
